client/HeroMovementController.cpp

```diff
@@ -0,0 +1,378 @@
+
+	auto prevTile = LOCPLINT->cb->getTile(posPrev);
+	auto nextTile = LOCPLINT->cb->getTile(posNext);
+
+	auto prevRoad = prevTile->roadType;
+	auto nextRoad = nextTile->roadType;
+	bool movingOnRoad = prevRoad->getId() != Road::NO_ROAD && nextRoad->getId() != Road::NO_ROAD;
+
+	if(movingOnRoad)
+		return nextTile->terType->horseSound;
+	else
+		return nextTile->terType->horseSoundPenalty;
+};
+
+void HeroMovementController::updateMovementSound(const CGHeroInstance * h, int3 posPrev, int3 nextCoord, EPathNodeAction moveType)
+{
+	// Start a new sound for the hero movement or let the existing one carry on.
+	AudioPath newSoundName = getMovementSoundFor(h, posPrev, nextCoord, moveType);
+
+	if(newSoundName != currentMovementSoundName)
+	{
+		currentMovementSoundName = newSoundName;
+
+		if(currentMovementSoundChannel != -1)
+			CCS->soundh->stopSound(currentMovementSoundChannel);
+
+		if(!currentMovementSoundName.empty())
+			currentMovementSoundChannel = CCS->soundh->playSound(currentMovementSoundName, -1, true);
+		else
+			currentMovementSoundChannel = -1;
+	}
+}
+
+void HeroMovementController::stopMovementSound()
+{
+	if(currentMovementSoundChannel != -1)
+		CCS->soundh->stopSound(currentMovementSoundChannel);
+	currentMovementSoundChannel = -1;
+	currentMovementSoundName = AudioPath();
+}
+
+bool HeroMovementController::canHeroStopAtNode(const CGPathNode & node) const
+{
+	if(node.layer != EPathfindingLayer::LAND && node.layer != EPathfindingLayer::SAIL)
+		return false;
+
+	if(node.accessible != EPathAccessibility::ACCESSIBLE)
+		return false;
+
+	return true;
+}
+
+void HeroMovementController::movementStartRequested(const CGHeroInstance * h, const CGPath & path)
+{
+	assert(duringMovement == false);
+	duringMovement = true;
+	currentlyMovingHero = h;
+
+	CCS->curh->hide();
+	moveHeroOnce(h, path);
+}
+
+void HeroMovementController::moveHeroOnce(const CGHeroInstance * h, const CGPath & path)
+{
+	// Moves hero once, sends request to server and immediately returns
+	// movement alongside paths will be done on receiving response from server
+
+	assert(duringMovement == true);
+
+	const auto & currNode = path.currNode();
+	const auto & nextNode = path.nextNode();
+
+	assert(nextNode.turns == 0);
+	assert(currNode.coord == h->visitablePos());
+
+	int3 nextCoord = h->convertFromVisitablePos(nextNode.coord);
+
+	if(nextNode.isTeleportAction())
+	{
+		stopMovementSound();
+		logGlobal->trace("Requesting hero teleportation to %s", nextNode.coord.toString());
+		LOCPLINT->cb->moveHero(h, nextCoord, false);
+		return;
+	}
+	else
+	{
+		updateMovementSound(h, currNode.coord, nextNode.coord, nextNode.action);
+
+		assert(h->pos.z == nextNode.coord.z); // Z should change only if it's movement via teleporter and in this case this code shouldn't be executed at all
+
+		logGlobal->trace("Requesting hero movement to %s", nextNode.coord.toString());
+
+		bool useTransit = nextNode.layer == EPathfindingLayer::AIR || nextNode.layer == EPathfindingLayer::WATER;
```

Comment:
What if monster is located on water? Or event? Will it be ignored?

Reply:
Huh, this is actually bug in H3.
In H3, if you use water walk and trigger event hero will stop on water. Looks to be fixed in HDMod+HotA

Will recheck vcmi.

Reply:
VCMI behavior (with this PR)

- Monster + fly = not targetable in UI, hero walks on top of monster
- Monster + water walk = not targetable in UI, hero walks around guarded tiles
- Event + fly = ignored
- Event + water walk = ignored

Reply:
> event + water walk = ignored

What about boat? Do boats trigger events? I know there are maps with them

Reply:
Yes, they do. "transfer" mode is only used for AIR and WATER layers. Boats use SAIL layer.